### DISCUSSION TITLE
Problem in Output of ants Registration interfaced via Nipype : FileNotFoundError output

### DISCUSSION THREAD

**Discussion by BastiaanvanTol** [ANTsX/ANTs#1613] [Q&A]
Hello all,

I want to interface ANTs in Python using Nipype for cross-modal registration of MRI and PET images. I have compiled ANTs as suggested, and Nipype is working as well. I am able to import the Registration module via from `nipype.interfaces.ants import Registration`

However, when I define the transformation parameters and I try to run the registration, I have not been able to deal with a **FileNotFoundError** that keeps appearing. It says:

**_No such file or directory '/Users/bastiaanvantol/Desktop/Tau_PET/output_ants/output_Composite.h5' for output 'composite_transform' of a Registration interface_**

Can anybody help me where this problem may lie, and suggestions how I can fix this?
Any help would be greatly appreciated, thanks!

----------------------------------------------------------------------------------------------------------

Here's is the example code from the Nipype website that I use to try out interfacing:
```
reg = Registration()
reg.inputs.fixed_image = pet_img
reg.inputs.moving_image = mr_img
reg.inputs.output_transform_prefix = "/Users/bastiaanvantol/Desktop/Tau_PET/output_ants/output_"
reg.inputs.transforms = ['Affine', 'SyN']
reg.inputs.transform_parameters = [(2.0,), (0.25, 3.0, 0.0)]
reg.inputs.number_of_iterations = [[1500, 200], [100, 50, 30]]
reg.inputs.dimension = 3
reg.inputs.write_composite_transform = True
reg.inputs.collapse_output_transforms = False
reg.inputs.initialize_transforms_per_stage = False
reg.inputs.metric = ['Mattes']*2
reg.inputs.metric_weight = [1]*2 # Default (value ignored currently by ANTs)
reg.inputs.radius_or_number_of_bins = [32]*2
reg.inputs.sampling_strategy = ['Random', None]
reg.inputs.sampling_percentage = [0.05, None]
reg.inputs.convergence_threshold = [1.e-8, 1.e-9]
reg.inputs.convergence_window_size = [20]*2
reg.inputs.smoothing_sigmas = [[1,0], [2,1,0]]
reg.inputs.sigma_units = ['vox'] * 2
reg.inputs.shrink_factors = [[2,1], [3,2,1]]
reg.inputs.use_estimate_learning_rate_once = [True, True]
reg.inputs.use_histogram_matching = [True, True] # This is the default
reg.inputs.output_warped_image = '/Users/bastiaanvantol/Desktop/Tau_PET/output_ants/warped_image.nii'
reg.cmdline
reg.run()  
```

**Comment by cookpa** [comment]
Can you please add all the output including reg.cmdline - so we can see what the ANTs command was

**Reply by BastiaanvanTol** [reply]
Hi, thank you for answer. I have figured out what the problem was, and it was relatively simple; one inaccuracy in referring to input directory (it did not read a single space in one folder name correctly.

Now I tried to run just a rigid registration from MR to PET in native space, but when I open the warped_output_image, the nifti-file loads in correctly, though it is completely off-centre: in every dimension(x, y, z), I only see half or maybe 30% of the brain.

Do you have any idea on what could be the cause of this?

**Reply by BastiaanvanTol** [reply]
The output nifti image looks like this, I will add it zipped for visualisation purposes as well:

<img width="617" alt="Screenshot 2023-10-21 at 21 46 17" src="https://github.com/ANTsX/ANTs/assets/124936795/2516cb77-46e4-4a9e-9ae2-a968f16a8084">

[warped_image.nii.zip](https://github.com/ANTsX/ANTs/files/13061701/warped_image.nii.zip)

**Reply by cookpa** [reply]
The inputs would be more informative than the output.

These kinds of issues are usually problems with the initial lack of overlap of physical space. The first thing `antsRegistration` does is to align the centers of mass of the images. If that puts them too far apart, the optimization cannot converge.

You can preview this interactively using ITK-SNAP's registration tool. Load the fixed image then add the moving image as an overlay, then go to registration and try to align COM.

**Reply by BastiaanvanTol** [reply]
Thank you Philip. Indeed the two images are far apart when I overlay them.

When I align them COM in ITK-SNAP, they fit very nicely on top of each other, thus solving the problem I had with my antsRegistration output. This is very good news! But it also means that ITK-SNAP did not have any problems with the COM alignment. Naturally I would like to have this step happen in my python pipeline as well, because I will be doing the coregistration for many images.

Is there a way for me to separately align the images COM with ANTs before I do the Rigid registration?

**Reply by cookpa** [reply]
Yes you can use `antsAI`

**Reply by BastiaanvanTol** [reply]
Thanks a lot for suggesting this function; I did not know I could use antsAI for this purpose, and it has been working well. Thank you!

**Comment by gdevenyi** [comment]
You didn't use an `initial_moving_transform` in your nipype call.

**Reply by BastiaanvanTol** [reply]
Thank you Gabriel! I defined `initial_moving_transform ` using the `trans.mat` file coming from antsAI, and that solved my issue.

**Reply by cookpa** [reply]
There's also an option `initial_moving_transform_com`, which will do the COM alignment for you without the need to call `antsAI`.

**Reply by BastiaanvanTol** [reply]
Aha I see. Even more simple. I also found that using registrationSyNQuick does COM correction automatically, and for both my rigid & rigid-affine-SyN transformations, it performs very well. So, I think for my purposes, this is working great with increased computational efficiency.

**Reply by cookpa** [reply]
That's correct, in `antsRegistration` the option is `-r [ fixed, moving, 1 ]` and is done in the ants scripts. If omitted, it is assumed that the physical spaces of the images already overlap.
